lib/Runtime/Library/JSON.cpp

```diff
@@ -647,42 +679,68 @@ namespace JSON
             }
             else
             {
-                uint32 precisePropertyCount = 0;
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    bool isPrecise;
-                    uint32 propertyCount = GetPropertyCount(object, &enumerator, &isPrecise);
-                    if (isPrecise)
-                    {
-                        precisePropertyCount = propertyCount;
-                    }
+                      Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
+                      if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                      {
+                        int propertyCount = dynamicObject->GetPropertyCount();
 
-                    if (ReplacerFunction != replacerType)
-                    {
-                        // Reserve initial slots for properties. +1 stands for extra property being pushed during the stringify
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount + 1);
+                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
+                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
                         enumerator.Reset();
-                        while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                        Js::PropertyId previousId;
+                        if (dynamicObject->HasObjectArray())
                         {
-                             if (id == Js::Constants::NoProperty)
+                            int totalNumPropertyCount = this->GetPropertyCount(object, &enumerator) - propertyCount;
+                            Assert(totalNumPropertyCount > 0);
```

Comment:
yes/no? Sorry, didn't understand?

Reply:
Assert won't do anything under test/release bits. We can use a failfast to die out if the property count is unexpected.

Reply:
well, I understood as if it happens in release mode, so the loop below goes into infinite loop, would we fail fast?.... 

The asserted scenario should not happen (if it's happening, something is off and we must see it on debug mode). If anyone expects it might happen??, then it's easy to replace `assert` with an `if` check.

Reply:
That's the intention of failfast. We might not see this in debug mode because of the lack of coverage compared to in-product coverage, so we shouldn't rely on it to catch this case. This is a common pattern in security issues.

AssertOrFailFast is better to use unless there's a specific way we want to recover here, or the branch has perf implications.

Reply:
@tcare or anyone else, from the implementation, do you know any tangible reason that we should explicitly check this? Otherwise, well, we could failfast so many unlikely stuff though.